### ISSUE TITLE
Windows (NTFS): Inconsistent support for the long-path prefix (\\?\) in filesystem paths

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#10805]
Note: This is a generalization of #4439. @SteveL-MSFT, if you agree, please close the latter.

Prefixing full, normalized, native filesystem paths with `\\?\` allows targeting filesystem items whose path is longer than the legacy limit of 259 characters.

As an aside: In Windows 10 you can now opt in system-wide to support long paths, but individual applications must _also_ opt in (PowerShell does) - see https://blogs.msdn.microsoft.com/jeremykuhne/2016/07/30/net-4-6-2-and-long-paths-on-windows-10/

As of PowerShell Core 7.0.0-preview.4, **`\\?\`-prefixed paths _are_ supported if you use the `-LiteralPath` parameter** (as opposed to (implied) `-Path`) with cmdlets such as `Get-ChildItem` and `Get-Item`.

However, **`\\?\`-prefixed paths are _not_ supported in the following scenarios**:

*  Using the `-Path` parameter (which is implied if omitted) with cmdlets such as `Get-ChildItem` and `Get-Item` (but, curiously, it does work with `Remove-Item`).
    * This is consistent with @jeffbi's comment at https://github.com/PowerShell/PowerShell/issues/4439#issuecomment-319578300, "t looks to me like this is more in the globbing code than in the provider itself." - i.e, the problem is related to interpreting the argument as a _wildcard_ expression.

*  Using `>` / `>>`, which implicitly behave like `Out-File -Path` - see #4726

* In the same vein, _invocation of an executable_ - wether with or without `&` - is broken (#4726 as well).

* `Start-Process`

* `New-Item`

Note: 

* Some of these problems are regressions from Windows PowerShell, where only the invocation / `Start-Process` tests fail and `>` only with a _new_ file.

* I haven't looked into whether invoking an _executable_ with an overly long path is supported in principle by the underlying APIs.

# Steps to reproduce

Run the following from a Pester test script (`*.Tests.ps1`) on Windows:

```powershell
Describe "Support for long paths, via \\?\" {
  BeforeAll {
    Push-Location (Convert-Path TestDrive:\)
    $dirPath = $PWD.ProviderPath
    $PrefixedDir = "\\?\$dirPath"
    $fileName = ('a' * 248 + '.cmd')
    $fileNameAlt = ('b' * 248)
    $PrefixedFullName = "$PrefixedDir\$fileName"
    $PrefixedFullNameAlt = "$PrefixedDir\$fileNameAlt"
    # Create the file with the overly long path using .NET,
    # to avoid issues with New-Item
    [IO.File]::WriteAllText($PrefixedFullName, '')
  }
  It "Get-ChildItem -LiteralPath" {
    Get-ChildItem -LiteralPath $PrefixedFullName | % FullName | Should -Be $PrefixedFullName
  }
  It "Get-ChildItem -Path" {
    Get-ChildItem -Path $PrefixedFullName | % FullName | Should -Be $PrefixedFullName
  }
  It "Remove-Item -LiteralPath" {
    { Remove-Item -LiteralPath $PrefixedFullName } | Should -Not -Throw
    # Recreate the file.
    [IO.File]::WriteAllText($PrefixedFullName, '')
  }
  It "Remove-Item -Path" {
    { Remove-Item -Path $PrefixedFullName } | Should -Not -Throw
    # Recreate the file, if necessary
    [IO.File]::WriteAllText($PrefixedFullName, '')
  }
  It "> with new file" {
    { '' >  $PrefixedFullNameAlt } | Should -Not -Throw
  }
  It "> / >> with existing file." {
    { '@echo Hi.' >  $PrefixedFullName } | Should -Not -Throw
    { 'REM ' >> $PrefixedFullName } | Should -Not -Throw
  }
  It "Invocation / &" {
    & $PrefixedFullName | Should -Be 'Hi.'
  }
  It "Start-Process" {
    Start-Process -FilePath $PrefixedFullName
  }
  It "New-Item" {
    { New-Item -Force -Type File $PrefixedFullName } | Should -Not -Throw
  }
  AfterAll {
    # Use .NET to remove the overly long path, so that Pester itself doesn't 
    # fail on trying to remove the dir. underlying TestDrive:
    [IO.File]::Delete($PrefixedFullName)
    [IO.File]::Delete($PrefixedFullNameAlt)
    Pop-Location
  }
}
```

# Expected behavior

All tests should pass.

# Actual behavior

All tests but the first one fail, with various error messages.

# Environment data

```none
PowerShell Core 7.0.0-preview.4
```

**Comment by iSazonov** [comment]
LocalGlobbing code is very sensitive and it seems it is not covered by tests explicitly - I think it is main stopper of a progress here.

**Comment by mklement0** [comment]
As of PowerShell Core 7.0.0-rc.2 `Get-Item -LiteralPath \\?\C:\Windows` is now broken too (complains about not finding the path; `Get-ChildItem -LiteralPath \\?\C:\Windows` still works).

I've noticed that the `\\?\` prefix is never _needed_ in PS Core. Still, people may use it in cross-edition scripts.

**Comment by iSazonov** [comment]
> I've noticed that the \\?\ prefix is never needed in PS Core.

Long paths is Windows feature, not PS Core.  It seems we have a test for long path support.

**Comment by mklement0** [comment]
Yes, it only applies _on Windows_, as hopefully clearly implied by the issue title. PS Core should support long paths there too, and support has degraded from Windows PowerShell, and even further since v7.0.0-preview.4

What I was trying to say: In PowerShell Core _on Windows_, you can seemingly access paths longer than 259 chars. even _without_ the `\\?\` prefix, irrespective of whether long-path support is turned on system-wide.

**Comment by mklement0** [comment]
> It seems we have a test for long path support.

You can run the test from the initial post; for me, 7 out of the 11 tests fail as of v7.0.0-rc.2

**Comment by iSazonov** [comment]
> What I was trying to say: In PowerShell Core on Windows, you can seemingly access paths longer than 259 chars. even without the \\?\ prefix, irrespective of whether long-path support is turned on system-wide.

I don't understand how PowerShell Core can work with long paths if the feature is not enabled on Windows 10.

**Comment by mklement0** [comment]
That is a fair question, @iSazonov - I haven't looked into the why and how, but you can try it yourself:

First, open an _elevated_ session and run the following:

```powershell
Set-ItemProperty HKLM:\SYSTEM\CurrentControlSet\Control\FileSystem LongPathsEnabled 0
```

Then, open a regular session and run the following:

```powershell
Set-Location; 'hi' > ('x'*(259 - "$pwd".length)); (Get-Item x*).FullName.Length; Remove-Item x*
```

You'll get no error message and `260` as the output, which is the longer-than-259-chars. path of the temporarily created file.

Running the same code in _Windows PowerShell_ fails.

**Comment by iSazonov** [comment]
@mklement0 Please share your Windows 10 version. I guess latest versions can have the feature enabled by default.

**Comment by mklement0** [comment]
It's `Microsoft Windows 10 Pro (64-bit; Version 1909, OS Build: 18363.592)`, but I don't think that is the issue, because the very same code also works on Windows _7_ (where system-wide support for long paths didn't even exist).

**Comment by iSazonov** [comment]
@mklement0 If you replace 259 in your test with more large value (what about 400?) you get an error.

**Comment by mklement0** [comment]
That is an unrelated constraint: even in long paths the _individual components_ must not be longer than 255 chars., as far as I know. You'd have to design the test differently to get longer paths.

**Comment by iSazonov** [comment]
Ah :-) my bad.
Get-Item really doesn't process long paths - it split path and works with leaf (the individual component < 256) but returns FileInfo object with FullName which is > 255 (266 chars). Perhaps just the fact confuses us.
Then I tried to Set-Location and create subdirectory but get error on Set-Location. So long paths do not works.

**Comment by mklement0** [comment]
I don't see the behavior you're seeing; try the following code, which creates a directory path that is itself longer than 259 chars., then a file in it, targets the directory with both `Get-Item` and `Get-ChildItem` and the file with `Get-Item`, changes to the directory with the overly long path with `Set-Location`, removes the file with the overly long path, then cleans up - it all seems to work fine, both on Windows 7 and Windows 10.

```powershell
Push-Location ($dir = mkdir -force ($env:TEMP + "\$pid\" + 'x' * 250)); $file = "$pwd\test.txt"; 'hi' > $file; "dir path length: $($dir.FullName.Length)`nfile path length: $($file.length)"; Get-Item $dir; '###'; Get-ChildItem $dir; '###'; Get-Item $file; '###'; Set-Location $dir -passthru; Pop-Location; Remove-Item $file; Remove-Item (Split-Path $dir) -Recurse
```

Output on my Windows 10 machine:

```none
dir path length: 288
file path length: 297


    Directory: C:\Users\jdoe\AppData\Local\Temp\5816

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----           1/24/2020  4:59 PM                xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
                                                  xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
###

    Directory: C:\Users\jdoe\AppData\Local\Temp\5816\xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---           1/24/2020  4:59 PM              4 test.txt
###
-a---           1/24/2020  4:59 PM              4 test.txt
###

Drive        : C
Provider     : Microsoft.PowerShell.Core\FileSystem
ProviderPath : C:\Users\jdoe\AppData\Local\Temp\5816\xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
               xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
Path         : C:\Users\jdoe\AppData\Local\Temp\5816\xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
               xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

**Comment by iSazonov** [comment]
@mklement0 I found why I don't understand you :-)
https://github.com/PowerShell/PowerShell/pull/3960/files#diff-52b1a915619c71b288b3f92f944924c4R1156-R1162

**Comment by mklement0** [comment]
😁 - thanks for digging deeper.

**Comment by SteveL-MSFT** [comment]
.NET Core itself supports long file paths implicitly, while .NET Framework requires app opt-in.  I get that users may want to use the `\\?\` legacy syntax, but it's not clear to me how impactful that is to motivate spending the time to fix this.

**Comment by mklement0** [comment]
Good to know that .NET Core as a whole supporting long paths - seemingly even without the system-wide  opt-in.

The code that @iSazonov linked to seems to no longer be part of the repo (except for the tests, which don't use the `\\?\` prefix).

>  I get that users may want to use the \\?\ legacy syntax

* They _may have done so in the past_, which means that old code with `\\?\` prefixes that would otherwise be compatible breaks.

* They may _have to going forward_, if they want to write cross-edition code.

Given that the current behavior is clearly a regression from Windows, I suggest either:

* fixing the behavior (which you seem disinclined to do)

* documenting the _breaking change_.

**Comment by iSazonov** [comment]
I am not sure we can fix this now. I think it _must_ necessarily be in FileSystem provider V2.

**Comment by mklement0** [comment]
Understood, @iSazonov. 

With the intent to fix this _eventually_, the categorization shifts from _breaking change_ to _(known) bug_, so please label this issue accordingly.

**Comment by iSazonov** [comment]
Meta bug label :-)

**Comment by pjaclark** [comment]
This should be listed as a Known Issue causing breaking changes until resolved.

**Comment by musm** [comment]
Is this related to why I can't create very long paths?
```
> mkdir ('x' * 290)
New-Item: The filename, directory name, or volume label syntax is incorrect ....
```

**Comment by mklement0** [comment]
@musm, no, you're seeing a fundamental NTFS limitation that applies even when long-path support is enabled: _individual file names_ must not be longer than `255` chars (as opposed to the length of a _path_ as a whole).

While you could argue that the error message _technically_ applies, it is certainly vague.

**Comment by Liturgist** [comment]
What is the state of this issue? I see that no one is assigned. Will this be a will-not-fix?

**Comment by iSazonov** [comment]
> What is the state of this issue?

Nice to have :-)

**Comment by iSazonov** [comment]
> but when you've got a server with millions of files on them and thousands of long paths, how are you supposed to automate anything?

It is my headache :-( If supporting `\\?\` is not so critical now since Windows 10 API now can support long paths OOB, PowerShell can not be used for processing large file volumes because of very bad performance and now it's even worse than before. See #14469 - the MSFT team does not acknowledge this and I have no explanation for it - only bewilderment.

**Comment by NoRegressionFromWindowsPowerShell** [comment]
**TL;DR :** 
Regression has happened because **Windows PowerShell embedded the manifest file whereas PowerShell does not.** Please embed the manifest in `pwsh.exe
`

Just because .NET Core / .NET 5 / .NET 6 / .NET 7 those runtimes themselves support `longPathAware` , doesn't mean all .NET applications which are built on top of those runtimes will support it automagically.  

Be it .NET or Completely Native, each and every `.exe` executables has to separately embed the manifest file.

![Screenshot 2022-02-22 063152](https://user-images.githubusercontent.com/100173066/155044462-da18d3c6-d3dd-4a34-82f5-1cdabec9e6b0.png)

![Screenshot 2022-02-22 063247](https://user-images.githubusercontent.com/100173066/155044470-4a0756c0-faca-4c82-8326-bf007c1f0480.png)

---
---
**Long Path works on windows 10 v1607 and upwards with _\*per executables opt-in.\*_**
Benefit : No need for `\\? or UNC Path Handling Complexities` Nonsense.

---
build the `pwsh.exe` with following embedded manifest file.

**Special Note :** This manifest file has to be embedded into each executables.
That means, If your App contains `SomeApp1.exe` and `SomeApp2.exe` , the manifest file has to be embedded into each executables.

File name must be `app.manifest`

```xml
<?xml version="1.0" encoding="utf-8"?>
<assembly manifestVersion="1.0" xmlns="urn:schemas-microsoft-com:asm.v1">
<application xmlns="urn:schemas-microsoft-com:asm.v3">
    <windowsSettings xmlns:ws2="http://schemas.microsoft.com/SMI/2016/WindowsSettings">
        <ws2:longPathAware>true</ws2:longPathAware>
    </windowsSettings>
</application>
</assembly>
```
---
Users who wish to use long path must Set Long Path Aware On system wide through Registry : `Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem` there make a DWORD key `LongPathsEnabled` and set the value to 1

**_Why System Wide Long Path is not on by default (?) according to a MS engineer_** : https://github.com/microsoft/WindowsAppSDK/issues/875#issuecomment-850551419


---

**Following APIs change behaviors after opt-in :**

CopyFileW, CopyFile2, CopyFileExW, CreateFileW, CreateFile2, CreateDirectoryW, CreateDirectoryExW GetCurrentDirectoryW, RemoveDirectoryW, SetCurrentDirectoryW, CreateHardLinkW, CreateSymbolicLinkW, DeleteFileW, FindFirstFileW, FindFirstFileExW, FindNextFileW, GetFileAttributesW, GetFileAttributesExW, SetFileAttributesW, GetFullPathNameW, GetLongPathNameW, MoveFileW, MoveFileExW, MoveFileWithProgressW, ReplaceFileW, SearchPathW, FindFirstFileNameW, FindNextFileNameW, FindFirstStreamW, FindNextStreamW, GetCompressedFileSizeW, GetFinalPathNameByHandleW.


---

Ref:


1. https://docs.microsoft.com/en-us/archive/blogs/jeremykuhne/net-4-6-2-and-long-paths-on-windows-10#enabling-win32-long-path-support

2. https://docs.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=cmd

**Comment by iSazonov** [comment]
@ghost The issue is about consistent support of `\\?\`, not long paths in whole. As for manifest, it force to turn on the feature _per application_. Now users can turn on the feature on OS level. I am not sure we want to force to turn on the feature since we don't cover code by tests and community doesn't vote for this actively. You could ask about adding the manifest in new issue.

**Comment by mkht** [comment]
I have a simple script that deletes all files in a particular folder. I have added the `\\?\` prefix to the path so that it will work correctly even if the full path of the file is longer than 260 characters. This script worked fine in Windows PowerShell 5.1.

```PowerShell
$TargetFolder = '\\?\C:\MyTemp\'
Get-ChildItem -LiteralPath $TargetFolder -Force | Remove-Item -Force -Recurse
```
As a result of running this script in PowerShell 7.2.1, PowerShell attempted to delete all files in the C drive and the system was destroyed.

Despite the fact that PowerShell 7 is advertised as being highly compatible with Windows PoweShell 5.1, this kind of behavior is ridiculous.
I don't know how hard it is, but hopefully it will be fixed to work consistently with Windows PowerShell 5.1.

**Comment by iSazonov** [comment]
@mkht Really PowerShell FileSystem Provider was never designed to support `\\?\` prefix. I hope we will support this but not as particular case, but as generic PowerShell Direct path.

**Comment by JazBInKC** [comment]
I just submitted a bug report for Set-ACL.  I'm using paths like:
![image](https://user-images.githubusercontent.com/96395010/166866507-fd47d52e-8135-45fb-80de-b185223586c7.png) in the creation of  new folders and the subsequent change of owner using Set-ACL.
I'm having to wait 3 minutes for Set-ACL to timeout with each new  folder.  No error message.
Thinking the lack of support for `\\?\` might be the issue.

**Comment by mkht** [comment]
PR #15873 that has been merged into `v7.3.0-preview.7` seems to have partially resolved this issue.

In the case of the `Get-ChidItem` command, we can use the long-path prefix with the `-LiteralPath` parameter, but cannot be used with the `-Path` parameter.

**Comment by maysara** [comment]
I have worked around this issue by running test-path -literalPath "\\?\C:\test" -ErrorAction:Stop and handling the exception. but now with a .Net framework update (4.8) this behaviour has changed again!

I'm unable to find a reliable way to detect if it is supported (the error shows up with out-file -literalpath even if it stopped with test-path)

**Comment by DarkLite1** [comment]
On PowerShell 7.4.1 using `-LiteralPath` works fine with the added string for long paths `\\?\` but when that code is wrapped within a job, it's no longer working. See #21102 for more details.

**Comment by alanwilter** [comment]
So, is it going to be solved or not? **PS7 does need to work with legacy "\\?\UNC\"**.

**Comment by Akarinnnnn** [comment]
Copy-Item doesn't support long path, too. These command works well in temp folder, but the anime folder with super-long path not working.
![longlongpath, with unicode](https://github.com/PowerShell/PowerShell/assets/43724908/5ceaa055-12c0-4b70-8d50-47926b5c3d7e)

**Comment by CharlotteAster** [comment]
You can try using a tool like LongPathTool. It helps manage and delete files or folders with long path issues easily.
